### DISCUSSION TITLE
Replacing curses with textual?

### DISCUSSION THREAD

**Discussion by frosencrantz** [saulpw/visidata#1412] [General]
Any thoughts on replacing curses with https://github.com/Textualize/textual ?

Will McGugan has been doing a lot of work on displaying tables:
- https://twitter.com/hashtag/TextualDataTable

**Comment by saulpw** [comment]
I appreciate all the work Will has been putting into Textual, and it looks like it will be useful for many projects.  However, Textual needs to own the application loop, and so does VisiData.  It would be a large amount of work to replace curses, we would lose a lot of control over the display and input, and we'd be dependent on a library that is not stable/mature yet.

What functionality of Textual data tables are you missing in VisiData? Or are you suggesting it in the same way that people propose that we rewrite VisiData in Rust?

**Comment by frosencrantz** [comment]
My interest in this case is a little beyond the latest shiny thing.  Curses itself is behind the latest capabilities you can use with recent terminal emulators.  OK that is shiny...  

There are several things I expect Textual could help with:
- Add support for modern terminal formatting (SGR mouse beyond column 132, curly/double/dotted/dashed underlines/etc, truecolor)  
  - I think there is a way to use truecolor with recent versions of Python Curses, but I think you still have the slot limit ???
- Handle variably wide and/or tall cells (see twitter thread from original comment)
   - I feel this has been a pain point in Visidata for a while.  Things are in much better shape, but cells need to all have the same height, which might not be how the data is formatted.
- Allow more advanced cell coloring (multiple colors beyond just fg/bg)
- Allow more advanced formatting of cell values using Rich library formatting.
- Maybe reduce the need to re-render cells multiple times ??? 
   - Context  (https://github.com/saulpw/visidata/issues/584#issuecomment-643563139)
- Builtin support to view markdown text. 
  - Not necessary but nice to have.  AJ shows a usable solution today: [article](https://github.com/saulpw/visidata/issues/584#issuecomment-643563139)
- Maybe more split window options, though this might not be something you desire.

I don't really know Textual very well.  Would it be possible for Visidata and Textual to share the application loop?  Can you hook into the application loop?

I agree it can be painful to use something that isn't stable/mature yet.  Also, they could lose funding in the future.  Currently, it does feel like Will and company will be responsive to issues that are found.

**Comment by frosencrantz** [comment]
A couple updates:
- Whether you can hook into the application loop, there is a request for better docs (https://github.com/Textualize/textual/issues/600) and some hints on what to do.  The restrictions might be too much.
- Textual also has a devtools console, which might help with debugging, which is difficult at times with curses.

**Comment by frosencrantz** [comment]
To be honest, @saulpw,   I don't know if this makes sense.   For data parsing, VisiData 100% relies on 3rd party parsers, but the UI is made via custom curses.   Textual appears to have nicely solved some difficult problems.

I do think that Textual would allow someone to quickly come up with a VisiData alternative that looks **pretty**, but might not have the same power.  That might provide for some competition for VisiData.

**Reply by saulpw** [reply]
@frosencrantz I still haven't heard your response to this question:

> What functionality of Textual data tables are you missing in VisiData? Or are you suggesting it in the same way that people propose that we rewrite VisiData in Rust?

In other words, what functionality would VisiData gain by using Textual?  You're basically asking me to do weeks of work to redo VisiData's display framework, and subsequently more weeks of work when it has performance regressions, and I have not heard one reason to actually do it, except that Textual is popular and "pretty".  I don't know what other "difficult problems" you are referring to.  It is also **not mature**, and the API is in flux (the main work is on a long-lived branch that has not been merged yet).  It might be nice for new projects (and I'm thinking of trying to use it for a text game idea I have), but for VisiData, it is not something I am interested in doing right now.   With the same effort, I would be much more inclined to create a web-based interface for VisiData in order to reach a 100x larger audience.

This may change in the future as the API matures and I understand what there is to be gained (and lost!) by using it, but that is a much longer-term proposal (at least a few years away).  

On this topic and several others, I would love to schedule a chat with you, where we can have a synchronous discussion and we can both feel heard.  Please email me if you're interested in talking about this further.  (and I don't mean to stifle discussion, if anyone else wants to chime in, please feel free!)

**Comment by frosencrantz** [comment]
Sorry I don't have time tonight to really respond.  I tried to answer your question is it just shiny or is there something useful here with my earlier [comment].  I think all of the items in that bullet list have usefulness to some extent.  

For example, having rows of differing heights would allow VisiData to show rows with multiple lines of data, but not waste screen space for rows that just have one line of data.  This allows better packing of data on the screen.  As a user I might see more data.  Right now VisiData row height is per sheet not per row.  That would be a useful feature.

I agree Textual is just a pending piece of work, Will has tweeted a bunch of demos, but not actually released the code or documentation.  Much of what I talk about is based on viewing the demos, and not using/coding working code, so I could be wrong.  

Also, does it make sense for VisiData is also a separate question.  I think Textual might lower the bar for a competitive tool.

I also agree that having a web version could have a much wider impact than **any** changes to the terminal version of VisiData.  After first learning VisiData, I really wanted that interface everywhere I was viewing data on the web.

[comment]: https://github.com/saulpw/visidata/discussions/1412#discussioncomment-3027575.

**Reply by saulpw** [reply]
Sorry about missing your other comment today, I had :eyes: on it so clearly I'd seen it before.  Those are some good use cases, but to wit:

> For example, having rows of differing heights would allow VisiData to show rows with multiple lines of data, but not waste screen space for rows that just have one line of data.

I understand that this doesn't look pretty, and I don't like it myself, but having a dynamic row height was causing some really terrible behavior that I couldn't figure out how to reasonably address.  For instance, if you base the row height on the currently-displayed columns only, then when you scroll across columns, you will probably get completely different layouts, so the cursor jumps around with every column change, which feels incredibly disorienting.  And if you base it on all visible columns, then on sheets with large numbers of columns, the performance will be terrible (and you'll *still* have a lot of wasted space if the current columns on screen are not the columns with the multiline rows).  Not to mention the difficulty with making PageUp and PageDown inverses of each other, while still making sure that paging through rows isn't "skipping" rows (i.e. rows that will never show up on the screen).

The current solution of "row height per sheet" makes all these problems go away, at the cost of wasting screen space when multiline rows are enabled.  So in practice I press `v` to see lines for bigger columns, and quickly press `v` to go back to the 'normal' display, which is not perfect, but the best of all options I've found so far.  (I found multiline rows to be quite frustrating to implement well, and I almost eliminated the feature entirely, before settling on the current less-than-ideal solution).

Now, maybe Textual has "solved" this difficult problem (and @willmcgugan is a smart fellow), but I think it is quite likely that it has only solved it for a much smaller set of use cases, and it would still feel disorienting and/or have performance problems with larger datasets that VisiData currently supports adequately.

> I think Textual might lower the bar for a competitive tool.

I think this would be great!  I don't think VisiData should be the only tool that does what it does.  If making something like VisiData using Textual delivers superior results, then by all means, let's see it.  I would love to try it out myself sometimes.

**Reply by willmcgugan** [reply]
I'd love to see some kind of marriage of Visidata and Textual. I've always had that use-case in mind, especially when working on the data table.

I agree with @saulpw 's assessment that *right* now may not be the time. Tacking on a Textual UI to Visidata is likely to be a big job. Curses and Textual are two quite different APIs separated by decades and switching one for the other is closer to a re-write than a port.

Visidata also does a bunch of things that Textual doesn't yet. I haven't solved *all* of the data table issues for instance.

Still, I think there could be some exciting synergy (jeez I'm talking like a CEO) in the future. Saul, if you ever do want to pick up Textual I'd be happy to give you a tour. We're close to a new release with docs.

**Reply by saulpw** [reply]
Thanks for the offer, Will, I might just take you up on that!  I would love for VisiData to be able to use Textual's widgets, they seem easy to create and they do look really good.  It will probably be awhile before Textual supports all the things VisiData needs, so I was thinking about how we could have a more gradual transition.  Is it at all possible for Textual to run alongside curses, so VisiData could maintain its own control loop and call into Textual as a library to layout and draw its widgets?  Then we could use Textual for the outermost interface (responsive panels, dialog layout, maybe multiline text input), and continue drawing the actual sheets with the existing code (which I also do not want to rewrite).

I get that this is a pretty wide gap, but if there's a chance we could make a VLT (VisiData Lettuce and Textual) I'd love to give it a try!

**Reply by willmcgugan** [reply]
I'm not sure, but maybe! Give it a few weeks. The first version powered by CSS will land and we will have some nice docs. That's probably a good point to look in to it. 

Worth exploring. That VLT sounds pretty tasty!

**Reply by charris-msft** [reply]
If you folks don't mind - I'd love to be a fly on the wall for that conversation. 

(I'm also very intrigued by @saulpw's comment about a web-based version... I've been working with [Superset](https://superset.apache.org) and [Metabase](https://metabase.com) lately, but they are still lacking when it comes to Exploratory Data Analysis (EDA), which I do a lot because of how little I know.)
